multipath-tools 0.10.1 and 0.10.2

mwilck opened the pull request:
## multipath-tools 0.10.2, 2025/02

This release contains backported bug fixes from the stable-0.11.y branch.

### Bug fixes

* Fix multipathd crash because of invalid path group index value, for example
  if an invalid path device was removed from a map.
  Fixes [#105](https://github.com/opensvc/multipath-tools/issues/105).
* Make sure maps are reloaded in the path checker loop after detecting an
  inconsistent or wrong kernel state (e.g. missing or falsely mapped path
  device). Wrongly mapped paths will be unmapped and released to the system.
  Fixes another issue reported in
  [#105](https://github.com/opensvc/multipath-tools/issues/105).
* Fix the problem that `group_by_tpg` might be disabled if one or more
  paths were offline during initial configuration.
* Fix possible misdetection of changed pathgroups in a map.
* Fix the problem that if a map was scheduled to be reloaded already,
  `max_sectors_kb` might not be set on a path device that
  was being added to a multipath map. This problem was introduced in 0.9.9.

## multipath-tools 0.10.1, 2025/01

This is the first bug fix release on the `stable-0.10.y` branch. It contains
bug fixes from 0.11.0, and some CI-related fixes.

### Bug fixes

* Fixed the problem that multipathd wouldn't start on systems with certain types
  of device mapper devices, in particular devices with multiple DM targets.
  The problem was introduced in 0.10.0.
  Fixes [#102](https://github.com/opensvc/multipath-tools/issues/102).
* Fixed a corner case in the udev rules which could cause a device not to be
  activated during boot if a cold plug uevent is processed for a previously
  not configured multipath map while this map was suspended. This problem existed
  since 0.9.8.
* Fixed the problem that devices with `no_path_retry fail` and no setting
  for `dev_loss_tmo` might get the `dev_loss_tmo` set to 0, causing the
  device to be deleted immediately in the event of a transport disruption.
  This bug was introduced in 0.9.6.
* Fixed the problem that, if there were multiple maps with deferred failback
  (`failback` value > 0 in `multipath.conf`), some maps might fail back later
  than configured. The problem existed since 0.9.6.
* Removed a warning message that multipathd would print if systemd's
  `WATCHDOG_USEC` environment variable had the value "0", which means that the
  watchdog is simply disabled. This (minor) problem existed since 0.4.9.
* Fixed a memory leak in the nvme foreign library. The bug existed since
  0.7.8.
* Fixed a problem in the marginal path detection algorithm that could cause
  the io error check for a recently failed path to be delayed. This bug
  existed since 0.7.4.

mwilck committed:
GitHub workflows: use {upload,download}-artifact@v4

The @v1 releases are deprecated.
https://github.blog/changelog/2024-02-13-deprecation-notice-v1-and-v2-of-the-artifact-actions/

Signed-off-by: Martin Wilck <mwilck@suse.com>

mwilck committed:
GitHub workflows: update dawidd6/action-download-artifact

Signed-off-by: Martin Wilck <mwilck@suse.com>

mwilck committed:
Github workflows: native.yml: use mosteo-actions/docker-run

We can't use "container:" any more because upload-artifact@v4 doesn't
work on Debian Jessie.

Signed-off-by: Martin Wilck <mwilck@suse.com>

mwilck committed:
GitHub workflows: native.yml: use extra job for clang

Running "make" in a container and "make clean" outside doesn't
work (access right issues). So just use separate jobs.

Signed-off-by: Martin Wilck <mwilck@suse.com>

mwilck committed:
GitHub workflows: native.yaml: make test and archive separately

Avoid "text file busy" error on GitHub.

dmevents-test: Text file busy
Makefile:74: recipe for target 'dmevents.out' failed

Signed-off-by: Martin Wilck <mwilck@suse.com>

bmarzins committed:
multipathd: fix deferred_failback_tick for reload removes

If reload_and_sync_map() removes the multipath device,
deferred_failback_tick() needs to decrement the counter so that it
doesn't skip the following device.

Reviewed-by: Martin Wilck <mwilck@suse.com>
Signed-off-by: Benjamin Marzinski <bmarzins@redhat.com>

mwilck committed:
11-dm-mpath.rules.in: import DM_COLDPLUG_SUSPENDED only once

We import DM_COLDPLUG_SUSPENDED in all code flows below mpath_coldplug_end.
Clarify this in the code.

Signed-off-by: Martin Wilck <mwilck@suse.com>
Reviewed-by: Benjamin Marzinski <bmarzins@redhat.com>

mwilck committed:
11-dm-mpath.rules.in: handle inactive suspended devices correctly

Since b22c273 ("11-dm-mpath.rules: Don't force activation while device is
suspended"), we've handled the case where a device is suspended while
an uevent is processed (e.g. because multipathd is reloading the
map again at the same time). But we were missing the case where
The device had never been initialized before. If .MPATH_DEVICE_READY_OLD
was empty, we'd jump to scan_import without setting MPATH_DEVICE_READY
to 0. This can cause a device not to be fully activated at boot time,
because in follow-up uevents we'd assume that the device had already
been set up.

Treat the case in which an uevent is processed for a previously not
fully set-up, suspended device like other situations where we set
MPATH_DEVICE_READY to 0.

Fixes: b22c273 ("11-dm-mpath.rules: Don't force activation while device is
suspended")
Reviewed-by: Benjamin Marzinski <bmarzins@redhat.com>

mwilck committed:
11-dm-mpath.rules.in: clarify DM_ACTIVATION logic

Our code is always setting MPATH_UNCHANGED and DM_ACTIVATION in
pairs. While DM_ACTIVATION is a global DM property, MPATH_UNCHANGED
is owned by us. Just set MPATH_UNCHANGED, and adapt DM_ACTIVATION
when necessary just in one place.

Signed-off-by: Martin Wilck <mwilck@suse.com>
Reviewed-by: Benjamin Marzinski <bmarzins@redhat.com>

mwilck committed:
11-dm-mpath-rules.in: skip one .DM_NOSCAN check

We set .DM_NOSCAN above where we set DM_UDEV_DISABLE_OTHER_RULES_FLAG, too. If
the latter isn't set, .DM_NOSCAN can't be set. Skip the redundant test.

Signed-off-by: Martin Wilck <mwilck@suse.com>
Reviewed-by: Benjamin Marzinski <bmarzins@redhat.com>

mwilck committed:
11-dm-mpath.rules.in: set .DM_NOSCAN if MPATH_UNCHANGED is set

When multipath reloads a device or fails or restores a path, the udev
rules disable LVM scanning, but since .DM_NOSCAN isn't set, blkid is
still run on the device. When multipath devices that are set to
queue_if_no_path lose all their paths at close to the same time, udev
workers can hang trying to run blkid. The blkid results shouldn't
change when multipathd is adding, removing, failing or reinstating
paths, aside from avoiding hanging udev processes, we're skipping
unnecessary work.

Hence, set .DM_NOSCAN if MPATH_UNCHANGED is set, to avoid blkid from
being called in 13-dm.rules.

Suggested-by: Benjamin Marzinski <bmarzins@redhat.com>
Signed-off-by: Martin Wilck <mwilck@suse.com>
Reviewed-by: Benjamin Marzinski <bmarzins@redhat.com>

mwilck committed:
libmpathutil: avoid -Wcast-function-type-mismatch error with clang 19

Avoid the following error with clang 19:

msort.c:268:27: error: cast from '__compar_fn_t' (aka 'int (*)(const void *, const void *)') to '__compar_d_fn_t' (aka 'int (*)(const void *, const void *, void *)') converts to incompatible function type [-Werror,-Wcast-function-type-mismatch]
  268 |         return msort_r (b, n, s, (__compar_d_fn_t)cmp, NULL);
      |                                  ^~~~~~~~~~~~~~~~~~~~

Signed-off-by: Martin Wilck <mwilck@suse.com>

mwilck committed:
libmultipath: don't set dev_loss_tmo to 0 for NO_PATH_RETRY_FAIL

If pp->dev_loss is DEV_LOSS_TMO_UNSET and min_dev_loss is 0 (which is
the case if no_path_retry is NO_PATH_RETRY_FAIL or NO_PATH_RETRY_UNDEF),
we will set pp->dev_loss to 0, which is wrong. Fix it.

Fixes: 058b5f5 ("libmultipath: fix dev_loss_tmo even if not set in configuration")
Signed-off-by: Martin Wilck <mwilck@suse.com>
Reviewed-by: Benjamin Marzinski <bmarzins@redhat.com>

mwilck committed:
libmultipath: dm_get_maps(): don't bail out for single-map failures

dm_get_maps() traverses the entire list of dm maps. We shouldn't
give up just because probing a single map failed.

Fixes: bf3a4ad ("libmultipath: simplify dm_get_maps()")
Signed-off-by: Martin Wilck <mwilck@suse.com>

mwilck committed:
multipathd: fix an unsigned int ovwerflow

Reported by coverity: "i--" may cause an underflow, which will again
cause an overflow when the loop continues. Use a signed int for
loops like this to make coverity happy.

Signed-off-by: Martin Wilck <mwilck@suse.com>

mwilck committed:
libmultipath: don't print error message if WATCHDOG_USEC is 0

WATCHDOG_USEC may be set to 0, which means that the watchdog
is disabled in systemd.

Fixes: 9366cfb ("multipathd: Implement systemd watchdog integration")
Signed-off-by: Martin Wilck <mwilck@suse.com>
Reviewed-by: Benjamin Marzinski <bmarzins@redhat.com>

mwilck committed:
GitHub workflows: enable unit tests for stable branches

Signed-off-by: Martin Wilck <mwilck@suse.com>

mwilck committed:
GitHub Workflows: add abi check for stable branches

The ABI should never change on a stable branch. This workflow
asserts that.

Signed-off-by: Martin Wilck <mwilck@suse.com>

mwilck committed:
Update NEWS.md for 0.10.1

Signed-off-by: Martin Wilck <mwilck@suse.com>

mwilck commented:
Just a demo for @bmarzins and @cvaroqui.

In reality, this would be a PR against a `stable-0.10.y` branch on opensvc/multipath-tools. The failing "check-abi" workflow wouldn't be run in that case.

bmarzins commented:
Looks good to me.

bmarzins commented:
Actually, I do have some thoughts. First, I'm not sure we need to update News.md for the stable branch releases. You're certainly welcome to, and I can see how it would save distro maintainers some time if they wanted an overview of the changes. But I think simply bumping the version number for each pull request should be all that's really necessary, if you want to skip some extra work.

Also, looking at this release makes it obvious that it's not always going to be easy to come up with a good "Fixes" commit trailer for some of the fixes. Without these, it won't be as easy as I hoped for people to be able to check if fix is applicable to a previous stable branch. Although I expect that if a fix is only applicable in the latest stable branch, then it should be pretty easy to come up
with a commit to use for a "Fixes:" trailer.

mwilck commented:
Updating NEWS.md is a minor effort, because I just copy/paste the relevant parts from the major release. The fact that I did it here doesn't mean I am committing myself to do it always in the future, but I think it wouldn't cost me much time.

About the "Fixes" trailers, I am not sure I understand your remark. Did you expect me to add "Fixes:" trailers to the bug fix commits that I'd identified? I opted to cherry-pick the 0.11.0 commits unchanged. IMO, if I add "Fixes:" here, I should do it in 0.11.0 as well (meaning that I need to rebase my "queue" branch, but I guess I'll have to do that anyway when I include the currently pending fixes).

In the future, we should agree to add "Fixes:" trailers to bug fix patches sent to dm-devel in the first place, but we haven't done so in the past (at least these trailers haven't been mandatory, we've been adding them rather sporadically).

Or were you referring to the other commits (like the github workflow commits) that don't fix actual bugs at all?

bmarzins commented:
> About the "Fixes" trailes, I am not sure I understand your remark. Did you expect me to add "Fixes:" trailers to the bug fix commits that I'd identified? I opted to cherry-pick the 0.11.0 commits unchanged. IMO, if I add "Fixes:" here, I should do it in 0.11.0 as well (meaning that I need to rebase my "queue" branch, but I guess I'll have to do that anyway when I include the currently pending fixes).
 
No. I was expecting just what you did. We shouldn't be unnecessarily mucking with commit messages for the stable branch. I was just noticing that a number of the actual bug fixes didn't include "Fixes" trailers, and for some of them it's not even obvious what you would put there. They are (or at least could be) fixing issues that were introduced in multiple commits or issues that have existed long enough that the code has been refactored multiple times since the bug was introduced. The code may have originally been correct, and then you would need to track down the code change that caused it to stop being the correct thing to do. 

I was hoping that downstream users who wanted to stick with a stable branch that we are no longer maintaining could just look through all of subsequent stable branches and pull out the fixes applicable to their branch. The easiest way to do that would be to look at the Fixes trailers, and see if their branch includes the referenced commit.

But like I said, it's not always obvious what you would put there.  I agree that we should probably make a better effort to label bugfixes with a Fixes trailer when they are initially committed, but it can probably be a best effort kind of thing. If the bug has existed for years and the code has been tweaked and refactored multiple times since the initial bug was introduced, I don't think we should require git spelunking to find the initial breaking commit. But in that case, the bug has been around for years so finding the initial commit isn't that important. I suppose we could helpfully add something
like, for instance:

`Fixes: pre-0.6.4`

To indicate that the bug was been there since at least 0.6.4. For commits that fix bugs introduced in multiple commits, we can probably just get by with a best effort to find the oldest breaking commit and mention that there are others.
For example, with your commit c62254bc ("11-dm-mpath.rules.in: clarify DM_ACTIVATION logic") we could probably just go with something like:

`Fixes: 20cd0df96 ("multipath/kpartx rules: avoid superfluous scanning") and others`

> Or were you refering to the other commits (like the github workflow commits) that don't fix actual bugs at all?

Nope. Things like those don't need any trailers.

mwilck commented:
@bmarzins discussion about future improvements and fixes tags aside, am I assuming correctly that you would ack this PR if I made it on opensvc?


> I was hoping that downstream users who wanted to stick with a stable branch that we are no longer maintaining could just look through all of subsequent stable branches and pull out the fixes applicable to their branch.

Note that since 0.10.0, I have tried to point out in NEWS.md in which previous version a given bug had been introduced. I did not put the commit IDs there though because NEWS.md should also be readable by non-technical users.

I haven't spent this effort for earlier releases, but still the "Bug fixes" section in NEWS.md should help downstream users.

bmarzins commented:
> @bmarzins discussion about future improvements and fixes tags aside, am I assuming correctly that you would ack this PR if I made it on opensvc?

Yeah. My thoughts were just thoughts, not objections to the PR.

mwilck commented:
I've pushed [the bugfix part](https://lore.kernel.org/dm-devel/20241114144224.248606-2-mwilck@suse.com/) of my systemd watchdog patch. I hope this is fine for you. The 0.11.0 solution is not suitable for the stable branch IMO.

mwilck committed:
libmultipath: fix handling of pp->pgindex

pp->pgindex is set in disassemble_map() when a map is parsed.
There are various possiblities for this index to become invalid.
pp->pgindex is only used in enable_group() and followover_should_fallback(),
and both callers take no action if it is 0, which is the right
thing to do if we don't know the path's pathgroup.

Make sure pp->pgindex is reset to 0 in various places:
- when it's orphaned,
- before (re)grouping paths,
- when we detect a bad mpp assignment in update_pathvec_from_dm().
- when a pathgroup is deleted in update_pathvec_from_dm(). In this
  case, pgindex needs to be invalidated for all paths in all pathgroups
  after the one that was deleted.

The hunk in group_paths is mostly redundant with the hunk in free_pgvec(), but
because we're looping over pg->paths in the former and over pg->pgp in
the latter, I think it's better too play safe.

Fixes: 99db1bd ("[multipathd] re-enable disabled PG when at least one path is up")
Fixes: https://github.com/opensvc/multipath-tools/issues/105
Signed-off-by: Martin Wilck <mwilck@suse.com>
Reviewed-by: Benjamin Marzinski <bmarzins@redhat.com>
(cherry picked from commit cd912cffa2797a18c47426c816afa8eb2eae5b22)
(cherry picked from commit 714c20bebba6911255a527d937e7a62764ffe338)

mwilck committed:
libmultipath: make pgcmp detect if map is missing a path group

The previous algorithm didn't detect the case case where cpgp
contained a path that was not contained in pgp. Fix this.

Cherry-picked from d4b35f61cb75c6e9b289e56c98457fc04ce4835e
Fixes: 90773ba ("libmultipath: resolve hash collisions in pgcmp()")
Signed-off-by: Martin Wilck <mwilck@suse.com>

mwilck committed:
libmultipath: reduce log level of "map X has multiple targets"

On systems with LVM volumes, "multipath -ll" will spit out lots of messages
like

    libmp_mapinfo: map vg-lv0 has multiple targets

which is irritating. Reduce the log level of these messages to 3, as they
are harmless most of the time.

This is a backport of e8949c2 ("libmultipath: reduce log level of
libmp_mapinfo() messages") from the master branch. We can't apply exactly
the same fix because the stable branch is missing 8c772d3 ("libmultipath:
check DM UUID earlier in libmp_mapinfo__").

Signed-off-by: Martin Wilck <mwilck@suse.com>

mwilck committed:
libmultipath: trigger uevents upon map creation in domap()

If map creation succeeds, previously not multipathed devices are
now multipathed. udev may not have noticed this yet, thus trigger
path uevents to make it aware of the situation. Likewise, if
creating a map fails, the paths in question were likely considered
multipath members by udev, too. They will now be marked as failed,
so trigger an event in this situation as well.

Fixes: https://github.com/opensvc/multipath-tools/issues/103
Suggested-by: Benjamin Marzinski <bmarzins@redhat.com>
Signed-off-by: Martin Wilck <mwilck@suse.com>
Reviewed-by: Benjamin Marzinski <bmarzins@redhat.com>
(cherry picked from commit 98b3a7bd9ed2a89f068fe40c253843cf78261905)

mwilck committed:
multipathd: trigger uevents upon map removal in coalesce_maps()

... if a map has been flushed. In this case, we know that the
the paths haven't been multipathed by coalesce_paths() because of the current
configuration (failure to create the map can't be the reason if the map
exists in coalesce_maps()). Make sure udev sees the paths which have
been released from the map as non-multipath.

Note that this is the only case where maps are flushed where it is correct
to trigger paths uevents. In other cases, e.g. after a "remove map" CLI
command, the configuration is unchanged and if we triggered an uevent,
the map would be re-created by multipathd when the uevent arrived.

Signed-off-by: Martin Wilck <mwilck@suse.com>
Reviewed-by: Benjamin Marzinski <bmarzins@redhat.com>
(cherry picked from commit ad3ea472b587c0c20d2100331b5c66f8602f8414)

bmarzins committed:
libmultipath: Don't skip set_path_max_sectors_kb()

If a multipath device already has need_reload set when a path is
adopted, it won't call set_path_max_sectors_kb() because of
short-circuit evaluation. This isn't what's intended.

Fixes: e5e20c7b ("libmultipath: set max_sectors_kb in adopt_paths()")
Signed-off-by: Benjamin Marzinski <bmarzins@redhat.com>
Reviewed-by: Martin Wilck <mwilck@suse.com>
(cherry picked from commit f5c0c4b25c5eaac87b78e6e0c8d52b0828c29893)

bmarzins committed:
libmultipath/foreign: fix memory leak in nvme foreign handler

_find_controllers() needs to free the udev device if it doesn't get
added to a path. Otherwise it can leak memory whenever check_foreign()
is called, causing multipathd's memory usage to continually grow.

Fixes: 7b47762 ("libmultipath: nvme: fix path detection for kernel 4.16")
Signed-off-by: Benjamin Marzinski <bmarzins@redhat.com>
Reviewed-by: Martin Wilck <mwilck@suse.com>
(cherry picked from commit 5a3d334e416a4a35ee88d7b8f1433ff7f57923ad)

chenrenhui-hw committed:
libmultipath: add condition for enqueueing path to io error check

In function io_err_stat_handle_pathfail(), path->io_err_dis_reinstate_time
is set to 0 to enqueue path to io error check as soon as possible. But
multipathd can not do it within marginal_path_err_recheck_gap_time seconds
after power-on, because curr_time is less than
marginal_path_err_recheck_gap_time.

To handle the early marginal path, we can enqueue path when
io_err_dis_reinstate_time is 0.

Signed-off-by: chenrenhui <chenrenhui1@huawei.com>
Reviewed-by: Benjamin Marzinski <bmarzins@redhat.com>
Reviewed-by: Martin Wilck <mwilck@suse.com>

>

(cherry picked from commit a1e3cf2d42cc4bab10753e466c8adb7efa83c99e)

mwilck committed:
Additional NEWS.md updates for 0.10.1

bmarzins committed:
libmultipath: stop static analyzer complaint in init_foreign

This change doesn't actually fix anything. The code was already safe.

Signed-off-by: Benjamin Marzinski <bmarzins@redhat.com>
Reviewed-by: Martin Wilck <mwilck@suse.com>
(cherry picked from commit 85ec51e7930b4cadfbc12718afa91ce3a4adf4b5)

bmarzins committed:
libmultipath: be lenient in allowing the alua-based pgpolicies

multipath wouldn't autodetect the GROUP_BY_PRIO path grouping policy or
allow the GROUP_BY_TPG policy if there was a path that didn't have its
prioritizer selected (for instance because multipathd was reconfigured
while it was offline). To avoid this, make verify_alua_prio() assume an
alua multipath device if all the paths with a prioritizer selected
(there must be at least one) use an alua-based prioritizer.

Signed-off-by: Benjamin Marzinski <bmarzins@redhat.com>
Reviewed-by: Martin Wilck <mwilck@suse.com>
(cherry picked from commit b47a577998eaff203018a00b57ba5e3674645848)

mwilck committed:
Update NEWS.md for 0.10.2

mwilck commented:
This PR is now broken because I've updated the master branch to the upstream master branch. It should be OK once I convert it into a PR to opensvc/stable-0.10.y, though.

mwilck committed:
GitHub Workflows: fix abi-stable.yaml for pull request

Signed-off-by: Martin Wilck <mwilck@suse.com>

mwilck commented:
This is now https://github.com/opensvc/multipath-tools/pull/109
